app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts

```diff
@@ -44,29 +175,73 @@ export default class LocalBackend extends Backend {
     }
   }
 
+  /** Return the ID of the root directory. */
+  override rootDirectoryId(): backend.DirectoryId | null {
+    return newDirectoryId(this.defaultRootDirectory)
+  }
+
   /** Return a list of assets in a directory.
    * @throws An error if the JSON-RPC call fails. */
-  override async listDirectory(): Promise<backend.AnyAsset[]> {
-    const result = await this.projectManager.listProjects({})
-    return result.projects.map(project => ({
-      type: backend.AssetType.project,
-      id: project.id,
-      title: project.name,
-      modifiedAt: project.lastOpened ?? project.created,
-      parentId: backend.DirectoryId(''),
-      permissions: [],
-      projectState: {
-        type: LocalBackend.currentlyOpenProjects.has(project.id)
-          ? backend.ProjectState.opened
-          : project.id === LocalBackend.currentlyOpeningProjectId
-            ? backend.ProjectState.openInProgress
-            : backend.ProjectState.closed,
-        // eslint-disable-next-line @typescript-eslint/naming-convention
-        volume_id: '',
-      },
-      labels: [],
-      description: null,
-    }))
+  override async listDirectory(
+    query: backend.ListDirectoryRequestParams
+  ): Promise<backend.AnyAsset[]> {
+    const parentIdRaw = query.parentId == null ? null : extractTypeAndId(query.parentId).id
+    const parentId = query.parentId ?? newDirectoryId(this.defaultRootDirectory)
+    const entries = await this.projectManagerFilesystemList(parentIdRaw)
+    return entries
+      .map(entry => {
+        switch (entry.type) {
+          case FileSystemEntryType.DirectoryEntry: {
+            return {
+              type: backend.AssetType.directory,
+              id: newDirectoryId(entry.path),
+              modifiedAt: entry.attributes.lastModifiedTime,
+              parentId,
+              title: fileInfo.fileName(entry.path),
+              permissions: [],
+              projectState: null,
+              labels: [],
+              description: null,
+            } satisfies backend.DirectoryAsset
+          }
+          case FileSystemEntryType.ProjectEntry: {
+            return {
+              type: backend.AssetType.project,
+              id: newProjectId(entry.metadata.id),
+              title: entry.metadata.name,
+              modifiedAt: entry.metadata.lastOpened ?? entry.metadata.created,
+              parentId,
+              permissions: [],
+              projectState: {
+                type: LocalBackend.currentlyOpenProjects.has(entry.metadata.id)
+                  ? backend.ProjectState.opened
+                  : entry.metadata.id === LocalBackend.currentlyOpeningProjectId
+                    ? backend.ProjectState.openInProgress
+                    : backend.ProjectState.closed,
```

Comment:
It seems like having a `LocalBackend.currentlyOpenProjects` results in two sources of truth for the state of the projects. Why not have `LocalBackend.projects` and let the projects take care of their own state?

Reply:
this awkward structure is due to only the `openProject` endpoint being the only one with enough info to construct the full project info...

... but i don't see any downsides of combining the two properties anyway, so i guess i'll do it

Reply:
If you want, you can open an issue describing which backend endpoints you'd like modified to provide the necessary info. The backend's job is to make the frontend's job easier ;)